### PR TITLE
Fill quick inserter results with available variations if limit is set and items are less.

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
There was a refactoring (https://github.com/WordPress/gutenberg/pull/24090)  in `embed block` that made it a single block with lots of `block variations`. That surfaced a problem with quick Inserter and the main Inserter when suggesting items with a limit (currently six).

A small fix was there to handle the cases where the `limit was set` and `the items where NOT fewer` than the limit.

I saw now that when you're searching for items and the filtered results are for example four, it continues to add all the available variations.

This PR handles this case ( `limit is set and bigger than items` ) and fills the Block list with available variations, if they exist, until it reaches the provided limit.

The addition now happens by taking account only the order of variations and there is not a sophisticated handling of which variations to choose. It could/should be handled better when we decide how to handle block variations better. 

## Steps to reproduce
1. Go to edit a post/page
2. Clock on the quick Inserter ( plus button )
3. Type in search `em`
4. You can see that the results are way more than the current limit ( 6 )

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by mcsf** [comment]
@ntsekouras: following our convo on Slack, I pushed a couple of commits to extend test coverage to **default block variations** and to ensure consistent business logic in `includeVariationsInInserterItems`. Let me know what you think.

**Comment by ntsekouras** [comment]
Thanks @mcsf! Great simplification 👍 .

I think we are just missing a case for prioritising the base blocks over variations. I tried quickly to handle it but thought only about the case when items length was equal with the limit ( check the test ), but I can now see that my thought wasn't implemented properly in case the items are less than the limit.

Will take a closer look tomorrow :)

File: packages/block-editor/src/components/inserter/utils.js [hunk 1]
```diff
@@ -29,6 +29,11 @@ const getItemFromVariation = ( item ) => ( variation ) => ( {
  * @return {Array} Normalized inserter items.
  */
 export function includeVariationsInInserterItems( items, limit = Infinity ) {
+	if ( items.length >= limit ) {
+		// No need to iterate for variations
+		return items.slice( 0, limit );
+	}
+
```
Comment (mcsf) [inline packages/block-editor/src/components/inserter/utils.js]: I think this is incorrect, and this patch to the first unit test shows why:

```diff
diff --git a/packages/block-editor/src/components/inserter/test/utils.js b/packages/block-editor/src/components/inserter/test/utils.js
index d09376f55f..94bb7ee6af 100644
--- a/packages/block-editor/src/components/inserter/test/utils.js
+++ b/packages/block-editor/src/components/inserter/test/utils.js
@@ -22,6 +22,16 @@ describe( 'inserter utils', () => {
 					id: 'core/block-with-default-variation-special',
 				} ),
 			] );
+			expect(
+				includeVariationsInInserterItems(
+					[ withDefaultVariationItem ],
+					1
+				)
+			).toEqual( [
+				expect.objectContaining( {
+					id: 'core/block-with-default-variation-special',
+				} ),
+			] );
 			// The base block type is supplemented by non-default variations
 			expect(
 				includeVariationsInInserterItems( [ withSingleVariationItem ] )
```

**Comment by mcsf** [comment]
> I just added again the quick exit if items are enough to avoid any iteration at all, given that this is called many times, for example with every search term change. Do you agree?
Of course we could also add there code to replace an item with their default variation, if exists.

Hm... my idea is that we shouldn't be trying to optimise without knowing what needs optimisation. That's because in this function it's clear that, the more straightforward and consistent it is, the more confidence we have in the results. Otherwise (as I [commented here](https://github.com/WordPress/gutenberg/pull/24481#discussion_r469235207)) when we optimise we give special treatments to certain values and forget to account for all the business rules.

In practice, even if we optimise, we will still need to loop over `items` to exclude types that have default variations. So that part can't be short-circuited. So the only thing left is the branching at `if ( filteredItems.length < limit ) {`, which doesn't cost anything. What else would you change?

> I think we are just missing a case for prioritising the base blocks over variations in case there are fewer or equal items with the limit and there are variations ( no default ) in a block before a block with default variation.

Is that still the case now that we have `return [ ...filteredItems, ...variationsToAdd ].slice( 0, limit );`? Maybe I didn't understand your point. If so, can you write a unit test?

**Comment by ntsekouras** [comment]
>In practice, even if we optimise, we will still need to loop over items to exclude types that have default variations. So that part can't be short-circuited.

>Hm... my idea is that we shouldn't be trying to optimise without knowing what needs optimisation.

I know that and I agree and that's why I don't know how much effort we should put in this right now. It should be handled with all the integration of `block variations` and find out what is needed and what will work best.

Even my suggestion for prioritizing  the `base blocks` would increase the complexity a lot without knowing if this will be the final direction for this.

I tried some work on prioritzing yesterday but needed feedback whether we should implement this or not. Just for reference an example test is below. Failing test because it removes the `withDefaultVariationItem` and when starts the loop adds the first variation of `withSingleVariationItem`.
```
               it( 'should prioritize base block elements', () => {
			const items = [ withSingleVariationItem, withDefaultVariationItem ];
			const res = includeVariationsInInserterItems( items, 2 );
			expect( res ).toEqual( [
				withSingleVariationItem,
				expect.objectContaining( {
					// id: 'core/embed-youtube',
					id: 'core/block-with-default-variation-special', // I believe this would be the expected result?
				} ),
			] );
		} );
```

Maybe keep your version for now, remove the last addition of mine and rethink/revisit the `block variations`? For now in master if someone searches for `embed` will have more results than expected.

What do you think?

**Comment by mcsf** [comment]
Thanks for your exploring and quick reply!

> Maybe keep your version for now, remove the last addition of mine and rethink/revisit the `block variations`? For now in master if someone searches for `embed` will have more results than expected.

Yeah, I agree with keeping it simple and not investing too much in this. Let's fix `master`.

**Review by mcsf — APPROVED** [review]
🚢

**Pull request merged**